n_helpers.c

```diff
@@ -74,9 +79,204 @@ static short normalYearDaysByMonth[] = {0, 31, 59, 90, 120, 151, 181, 212, 243,
+  @param  buffer A buffer to hold the result
+  @param  bufLen The total length of the buffer
+  @returns  NULL on success, else an error string pointer.
+  @note  Buffers are decoded in place, the original contents of the buffer will
+         be modified.
+*/
+/**************************************************************************/
+const char * NoteBinaryReceive(uint8_t * buffer, size_t bufLen)
+{
+    // Issue a "card.binary" request.
+    J *rsp = NoteRequestResponse(NoteNewRequest("card.binary"));
+    if (!rsp) {
+        return ERRSTR("unable to issue binary request", c_err);
+    }
+
+    // Ensure the transaction doesn't return an error
+    // to confirm the binary feature is available
+    if (NoteResponseError(rsp)) {
+        JDelete(rsp);
+        return ERRSTR("feature not implemented", c_bad);
+    }
+
+    // Examine "cobs" from the response to evaluate the space
+    // required to hold the COBS encoded data on the Notecard.
+    const size_t cobs = JGetInt(rsp,"cobs");
+    JDelete(rsp);
+    if (!cobs) {
+        return ERRSTR("no data on notecard", c_err);
+    }
+    if (cobs > bufLen) {
+        return ERRSTR("insufficient buffer size", c_err);
+    }
+
+    // Claim Notecard Mutex
+    _LockNote();
+
+    // Issue `card.binary.get` and capture `"status"` from response
+    char status[NOTE_MD5_HASH_STRING_SIZE] = {0};
+    J *req = NoteNewRequest("card.binary.get");
+    if (req) {
+        JAddIntToObject(req, "cobs", cobs);
+
+        // Ensure the transaction doesn't return an error.
+        J *rsp = NoteRequestResponse(req);
+        if NoteResponseError(rsp) {
+            JDelete(rsp);
+            return ERRSTR("failed to initialize binary transaction", c_err);
+        }
+
+        // Examine "status" from the response to evaluate the MD5 checksum.
+        strlcpy(status, JGetString(rsp,"status"), NOTE_MD5_HASH_STRING_SIZE);
+        JDelete(rsp);
+    }
+
+    // `NoteReceiveBytes()` of the COBS encoded bytes either in a single I/O of
+    // `cobsLen`, OR receive until newline and verify that you received
+    // `cobsLen` bytes, depending upon platform API
+    size_t overflow;
+    const char *errstr = _RawReceive(buffer, bufLen, false, 60000, &overflow);
+
+    // Release Notecard Mutex
+    _UnlockNote();
+
+    // Ensure transaction was successful
+    if (errstr) {
+        return ERRSTR(errstr, c_err);
+    }
+
+    // Check buffer overflow condition
+    if (overflow) {
+        return ERRSTR("unexpected data available", c_err);
+    }
+
+    // Decode it in-place which is safe because decoding shrinks
+    const size_t decLen = cobsDecode(buffer,bufLen,'\n',buffer);
+    buffer[decLen] = '\0';
+
+    // Verify MD5
+    char hash_string[NOTE_MD5_HASH_STRING_SIZE] = {0};
+    NoteMD5HashString(buffer, decLen, hash_string, NOTE_MD5_HASH_STRING_SIZE);
+    if (strncmp(hash_string, status, NOTE_MD5_HASH_STRING_SIZE)) {
+        return ERRSTR("data transfer error", c_err);
```

Comment:
I would change the error message to say the computed MD5 doesn't match the received MD5. No reason to not be specific about the failure, I think.